File: examples/nodejs-api/src/application/user/CreateUserUseCase.ts
// Application Service: CreateUserUseCase
import { User } from '../../domain/user/User';
import { UserId } from '../../domain/user/UserId';
import { Email } from '../../domain/user/Email';
import { UserRepository } from '../../domain/user/UserRepository';

export interface CreateUserRequest {
  id: string;
  email: string;
  name: string;
}

export class CreateUserUseCase {
  constructor(private readonly userRepository: UserRepository) {}
  
  async execute(request: CreateUserRequest): Promise<void> {
    const userId = UserId.create(request.id);
    const email = Email.create(request.email);
    
    // Check if email exists
    const existingUser = await this.userRepository.findByEmail(email);
    if (existingUser) {
      throw new Error('Email already registered');
    }
    
    const user = User.create(userId, email, request.name);
    await this.userRepository.save(user);
  }
}

File: examples/java-spring/src/main/java/com/example/demo/application/user/CreateUserUseCase.java
package com.example.demo.application.user;

import com.example.demo.domain.user.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Application Service
 * Orchestrates use cases, coordinates domain objects
 */
@Service
@Transactional
public class CreateUserUseCase {
    
    private final UserRepository userRepository;
    
    public CreateUserUseCase(UserRepository userRepository) {
        this.userRepository = userRepository;
    }
    
    public User execute(CreateUserCommand command) {
        Email email = Email.of(command.getEmail());
        
        if (userRepository.existsByEmail(email)) {
            throw new IllegalArgumentException("Email already exists");
        }
        
        User user = User.create(email, command.getName());
        return userRepository.save(user);
    }
}

// DTO for input
create class CreateUserCommand {
    private String email;
    private String name;
    
    // Getters and setters
    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
}
